File: fundamentos/desafio-js/src/questions/highest-balance-agency.ts
import { IAccount } from "../i-account";

/** A agência com maior saldo é a: */
export default function highestBalanceAgency(...accounts: Array<IAccount>) {
  const accs = new Map<number, number>();

  for (const acc of accounts) {
    accs.set(acc.agencia, (accs.get(acc.agencia) || 0) + acc.balance);
  }

  const acc = Array.from(accs.entries())
    .sort((a, b) => b[1] - a[1])[0];

  console.info(`A agência com maior saldo é a: ${acc[0]}, $${acc[1]}`);
}

File: fundamentos/desafio-dart/lib/questions/highest_balance_agency.dart
import 'package:desafio_dart/domains/account.dart';

/// A agência com maior saldo é a:
void highestBalanceAgency(List<Account> accounts) {
  var accs = Map<int, double>();

  for (var acc in accounts) {
    accs[acc.agencia] = ((accs[acc.agencia] ?? 0) + acc.balance);
  }

  final sorted = accs.entries.toList()
    ..sort((a, b) => b.value.compareTo(a.value));

  final acc = sorted.first;

  print('A agência com maior saldo é a: ${acc.key}, \$${acc.value}');
}
